perfmetrics/scripts/run_e2e_tests.sh

```diff
@@ -53,28 +53,162 @@ sudo apt-get install -y gcc python3-dev python3-setuptools
+    fi
+  done
+  return $test_fail_np_group_2
+}
+
+# Parallel execution of integration tests located within specified test directories.
+# It aims to improve testing speed by running tests concurrently, while providing basic error reporting.
+function run_parallel_tests() {
+  for test_dir_p in "${test_dir_parallel[@]}"
+  do
+    test_path_parallel="./tools/integration_tests/$test_dir_p"
+    # Executing integration tests
+    GODEBUG=asyncpreemptoff=1 go test $test_path_parallel -p 1 --integrationTest -v --testbucket=$BUCKET_NAME_PARALLEL --testInstalledPackage=$run_e2e_tests_on_package -timeout $INTEGRATION_TEST_TIMEOUT &
+    pid=$!  # Store the PID of the background process
+    pids+=("$pid")  # Optionally add the PID to an array for later
+  done
+
+  # Wait for processes and collect exit codes
+  for pid in "${pids[@]}"; do
+    wait $pid
+    exit_code_parallel=$?
+    if [ $exit_code_parallel != 0 ]; then
+      test_fail_p=$exit_code_parallel
+      echo "test fail in parallel: " $test_fail_p
+    fi
+  done
+  return $test_fail_p
+}
+
+# Test setup
+# Create Bucket for non parallel e2e tests
+# The bucket prefix for the random string
+bucketPrefix="gcsfuse-non-parallel-e2e-tests-group-1-"
+BUCKET_NAME=$(create_bucket $bucketPrefix)
+echo "Bucket name for non parallel tests: "$BUCKET_NAME
+BUCKET_NAME_NON_PARALLEL_GROUP_1=$BUCKET_NAME
+# Test directory array
+# These tests never become parallel as it is changing bucket permissions.
+test_dir_non_parallel_group_1=(
+  "readonly"
+  "managed_folders"
+)
+
+# Test setup
+# Create Bucket for non parallel e2e tests
+# The bucket prefix for the random string
+bucketPrefix="gcsfuse-non-parallel-e2e-tests-group-2-"
+echo "Bucket name for non parallel tests - 2 : "$BUCKET_NAME
+BUCKET_NAME=$(create_bucket $bucketPrefix)
+BUCKET_NAME_NON_PARALLEL_GROUP_2=$BUCKET_NAME
+# These test packages can be configured to run in parallel once they achieve
+# directory independence.
+# Test directory array
+test_dir_non_parallel_group_2=(
+  "explicit_dir"
+  "implicit_dir"
+  "operations"
+  "read_large_files"
+  "rename_dir_limit"
+)
+
+# Create Bucket for parallel e2e tests
+# The bucket prefix for the random string
+bucketPrefix="gcsfuse-parallel-e2e-tests-"
+BUCKET_NAME=$(create_bucket $bucketPrefix)
+echo "Bucket name for parallel tests: "$BUCKET_NAME
+BUCKET_NAME_PARALLEL=$BUCKET_NAME
+# Test directory array
+test_dir_parallel=(
+  "local_file"
+  "log_rotation"
+  "mounting"
+  "read_cache"
+  "gzip"
+  "write_large_files"
+  "list_large_dir"
+)
 
+# Run tests
+test_fail_p=0
+test_fail_np_group_1=0
+test_fail_np_group_2=0
 set +e
-# Executing integration tests
-GODEBUG=asyncpreemptoff=1 go test ./tools/integration_tests/... -p 1 --integrationTest -v --testbucket=$BUCKET_NAME --testInstalledPackage=$run_e2e_tests_on_package -timeout $INTEGRATION_TEST_TIMEOUT
-exit_code=$?
 
+echo "Running parallel tests..."
+# Run parallel tests
+run_parallel_tests &
+my_process_p=$!
+echo "Running non
```

Comment:
nit: you meant echo "Running non parallel tests group 1... "

Reply:
You have to review this PR 
https://github.com/GoogleCloudPlatform/gcsfuse/pull/1775/files

This is draft only on top of above PR.